### ISSUE TITLE
passing params to call on eventMacro only works for calls inside macros

### ISSUE THREAD

**Issue by Nipodemos** [OpenKore/openkore#905]
It's not exatly an issue, but it's odd.
In one of the last updates made in eventMacro, it is possible to pass params when you call macros, enabling a much better code reuse.

However, it's only possible when you call macros inside a macro block, and not possible when you call a macro with an automacro.
Example:

    automacro prepareForFishing {
    	InInventory "Meat" > 0
    	exclusive 1
    	call {
    	call fish "izlu2dun" "165 164"
    	}
    }

    maco fish {
        do move $.param1 $.param2 &rand ( 1,6 )
    	do talk &npc ( $.param2 )
    	do talk resp 0
    	do talk resp /Meat/
    }
its not possible to do something like that:

    automacro prepareForFishing {
        InInventory "Meat" > 0
    	exclusive 1
    	call fish "izlu2dun" "165 164"
    }

This for me is really odd, and make the code more confusing... 
So @allanon and @Henrybk , if one of you can see if it possible to make this better... i would be very thankful

**Comment by rhan4x** [comment]
![untitled](https://user-images.githubusercontent.com/30954281/29240818-e36e83ee-7f96-11e7-9f42-efdb662a7d6e.jpg)


Ayo gan pembeli yg di atas ajah udah ane bantu via teamviewer agan yg lain mana!!

Ayo gan sumber botnya langsung cuman di www.rhanfelta.blogspot.com think smart gan dari pada beli zeny mahal-mahal mending stock sendiri ajah gan bisa jualan zeny sendiri juga lagi, dari pada bayar jasa joki mahal cuman dpt 1 char mending pake bot bisa bikin char lvl 99 100ID++ dalam 1 pc dalam waktu singkat.

Ayo gan mumpung masih ada promo khusus buat agan 1x lagi. sebelum harga promonya expired mumpung murah gan ayoo join di www.rhanfelta.blogspot.com bot ragnarok bergaransi aman sumbernya langsung harga murah cuman buat agan.

update ane baru beli hp baru hp lama masih di ugd jadi udah bisa terima tlp sms lagi yg berminat contact ane ajah langsung.

081380644602 / visit www.rhanfelta.blogspot.com

**Comment by Zelecktor** [comment]
You have bad Concept of macros.
An automacro is just the params to call a macro, and the macro are all the actions that you want to do.
To make it works you should have something like this:

automacro prepareForFishing {
	InInventory "Meat" > 0
	overrideAI 1
	exclusive 1
	call fish
}

maco fish {
		$rand = @random("-2","-1","0","1","2")
		$map = izlu2dun #you can modify this
		$px = @eval(165+$rand)
		$py = @eval(164+$rand)
	do pause 1
	do move $map $px $py
	do talk npc (npc coordinates and steps) # example, do talk npc 80 30 c c r0 r1 n
	release prepareForFishing
}




So in that case your bot will not stand only in the same cell, it will choose a random square of 5x5 arroud the center 165 164

**Comment by Zelecktor** [comment]
Note when you use $. there are a list of variables on the macro plugin. $.map is your current map for example, it is very diferent to put $map. $map is what? you need to define it.

$thing = ??? you need to define it
$anyName = ??? you need to define it
$map = ??? define it

But if you use

$.thing = error, this variable is not on the macro plugin
$.pos = your current position
$.hp = current hp
$.map = the map you're on

**Comment by Nipodemos** [comment]
thanks for the reply, but i think you understood wrong what i want with these macros(probably because i didn't post the entire macro) , so let me explain:

```
automacro firstSpot {
	InInventory "Meat" > 0
	exclusive 1
        ConfigKeyNot spot 1
	call {
	call fish "izlu2dun" "157 63"
	}
}

automacro secondSpot {
	InInventory "Meat" > 0
	exclusive 1
	ConfigKey spot 1
	ConfigKeyNot spot 2
	call {
	call fish "prt_fild08" "321 307"
	}
}

automacro thirdSpot {
	InInventory "Meat" > 0
	exclusive 1
	ConfigKey spot 2
	ConfigKeyNot spot 3
	call {
	call pescar "alberta" "235 27"
	}
}

macro pescar {
	$rand = @random("-2","-1","0","1","2")
        $map = $.param1 
        $px = &eval($.param2 +$rand)
        $py = &eval($.param3 +$rand)
        do pause 1
        do move $map $px $py &rand
	do talk &npc ( $.param2 $param3 )
	do talk resp 0
	do talk resp /Meat/
}
```
the macro has more content, but it is irrelevant.
The point here, it's a better code reuse, simply passing custom params to every call
This macro is forn the fishing quest on re:start, and has other 12 spots, so making 1 macro for every automacro it's dumb, so i did this :-) 

Now i have some things to say to you @Zelecktor 
1- you know you cant put another value in the do move, that corresponds to the distance of the coordinates?? 
like: do move 164 165 &rand(1,6)
it's gonna move to that place, and pick a random distance from de coordinates between 1 and 6 cells.
that specific function makes this hole thing of using eval and random and rand pointless.
2- i think its better to use &rand() instead of &random() when the values are numbers
3- i know that variables that begin with a dot are special read-only variables. i've made a mistake related to this?
4-why a release on the end of macro?? it doesn't have any reason

That's all.
Thanks for the reply, it helped me to improve more my code

**Comment by Zelecktor** [comment]
Ohhh i see, i thought you need to do it specifically, not for a quest.
In that case i need to check it step by step because, i dont know how the quest works.
Release is to repeat the macro X times when the automacro params are on their conditions. Example if you run out meats = 0 then the macro will execute, then if you ran out again, then the macro will execute again.

To use the correct syntax for pre-designed variables we use "@", not "&", just in case. But to make like random stand spot, there are many ways to do it, and none is 100% correct, I wrote that method and worked fine for me, never got bugged.

Another thing is when you use an automacro with a macro inside it is like in this format

automacro blabla {
params
inventory
more params
call {
      all macros steps
      do this
      do another thing
      }
}

or you can use

automacro blabla {
params
inventory
more params
call boomboom
}

macro boomboom {
      all macros steps
      do this
      do another thing
      }
}

Both means exactly the same but in your macro i dont undertand this:

automacro firstSpot {
	InInventory "Meat" > 0     #this is correct, will execute when you got at least 1 meat
	exclusive 1                         # correct, it will run only 1 time
        ConfigKeyNot spot 1         # never seen before. see Conditions on the macro guide
	call {                                            #here put your macro
	call fish "izlu2dun" "157 63"       #here you are calling a macro called fish, openkore doesnt know what you want to do, but for example the macro name, must be the same. I know spanish so, you should put "pescar" because there is no macro called "fish". Another thing is even if you put it in that way, it wont work.
	}
}

A worked macro, should be like this:

automacro firstSpot {
      InInventory "Meat" > 0
      exclusive 1
      overrideAI 1    #use it, or your AI will let you move to another place, chance to get bugged
      call pescar
}

if you want to make it for each case you can do somethig like this

First make a line on your config.txt called:
StepFish Spot1

automacro SettingSpot1 {
    InInventory "Meat" > 0
    exclusive 1
    overrideAI 1
    eval $::config{StepFish} eq "Spot1"
    call Pescar
}

automacro Spot2 {
    InInventory "Meat" > 0
    exclusive 1
    overrideAI 1
    eval $::config{StepFish} eq "Spot2"
    call Pescar
}

automacro Spot3 {
    InInventory "Meat" > 0
    exclusive 1
    overrideAI 1
    eval $::config{StepFish} eq "Spot3"
    call Pescar
}

macro pescar {
                    $StepConfig = @eval($::config{StepFish})
                    if ($StepConfig == Spot1) goto Move1
                    if ($StepConfig == Spot2) goto Move2
                    if ($StepConfig == Spot3) goto Move3
   :Move1
   pause 1
   do move izlu2dun 157 163 #move near the npc
   pause 1
   do talknpc 165 164
   do talk resp 0
   do talk resp /Meat/  #you need to write the word "Meat"?, use do talk text Meat
   do conf Spot2
   release Spot2
   goto end

   :Move2
   pause 1
   do move prt_fild08 321 307 #move near the npc
   pause 1
   do talknpc 321 307
   do talk resp 0
   do talk resp /Meat/  #you need to write the word "Meat"?, use do talk text Meat
   do conf Spot3
   release Spot3
   goto end

   :Move3
   pause 1
   do move alberta 235 27 #move near the npc
   pause 1
   do talknpc 235 27
   do talk resp 0
   do talk resp /Meat/  #you need to write the word "Meat"?, use do talk text Meat
   goto end

   :end
   stop

}

**Comment by Nipodemos** [comment]
Now I understand
### Actually, i'm talking about **eventMacro plugin** , a rework of **macro plugin**.
you can look at your plugins folder, if it is updated, it has a eventMacro folder inside.
So that's why it seemed so odd and wrong to you what i've wrote.

And you right, i forgot to translate the macro "pescar" to "fish" =/ sorry 
(and i'm probably writing a lot of english words incorrectly)
```
automacro firstSpot {
InInventory "Meat" > 0 
exclusive 1 
ConfigKeyNot spot 1 # This condition checks if the key in config.txt "spot" has the value "1"
# exactly what your eval $::config{StepFish} eq "Spot1" does 
call { 
call pescar "izlu2dun" "157" "63" 
# the eventMacro accepts that you put params in front of the call of you macro
# each value inside "" is a param
# so izlu2dun is param1
# 157 is param2 
# 63 is param3
}
}
```
because of this params, is possible to write one single macro that any of the automacros can call
```
macro pescar {
	$rand = @random("-2","-1","0","1","2")
        $map = $.param1 # called by automacro first spot it this variable becomes izlu2dun
        $px = &eval($.param2 +$rand) # this one becomes 157 + the rand value
        $py = &eval($.param3 +$rand) # and this becomes 63   + the rand value
        #and when other automacro call , like secondspot, these values change and runs perfectly
        do pause 1
        do move $map $px $py &rand(1,4)
	do talk &npc ( $.param2 $param3 )
	do talk resp 0
	do talk resp /Meat/ #actually is to find an answer that have the word "meat" in it
        # it is easier than findind the correct number of the answer.
}
``` 
This macro you have made, it will work perfectly in **macro plugin**, but in **eventMacro plugin**, there is a better way (in my opinion) to write this.

Just a note, its more or less **15** steps the fishing quest, so in macro plugin i would have to write :Move1 , :Move2 and on until :Move15
that is too much 😟 
But, i thank you for all you work, i know you were trying to helping me, and i've noticed that you are pretty good in make macros, and you had a hard work to rewrite the macro for me. Thanks 😄 😃

**Comment by Zelecktor** [comment]
Ohhh i see, 15 steps is a lot. Well not all people uses eventMacro plugin. Macro plugin is like an universal language that you can share and it will works for all.
Yes, there are many ways to write a macro and all of them will get the same result, but it is better to use like "locked" variables and you got less chances to get bugged. Also, if your macro is shorter and simple is better than one larger.

I hope this give you an idea for this and future macros 😃

**Comment by zxc4xa** [comment]
Update Bukti 14 Agustus 2017 Setelah Mainatnce Still Works:
![14 agustus 2017 after mt](https://user-images.githubusercontent.com/30999621/29262785-46da9098-8100-11e7-836a-8df8028d6b08.jpg)

Beli Zeny Mahal? Jasa Joki per ID Muahaalll bangeet? mending join ajah gan di www.rhanfelta.blogspot.com lebih dari puluhan orang udah Join, Jadilah Smart Gamer gan dengan join di rhanfelta bisa stock zeny sendiri gak perlu beli lagi sama orang bahkan bisa jualan zeny sendiri. agan pun bisa bikin 100ID++ lvl99 dalam 1PC dengan waktu singkat come on gan think smart!!

buruan gan mumpung masih ada slot promo cuman untuk agan!!

Visit : www.rhanfelta.blogspot.com
Contact: 081380644602

**Comment by Nipodemos** [comment]
#1029
This pull request will solve this issue.
Just waiting to merge

**Comment by Nipodemos** [comment]
implemented #1029
